device_backends/include/AsyncDomain.h

```diff
@@ -0,0 +1,36 @@
+// SPDX-FileCopyrightText: Deutsches Elektronen-Synchrotron DESY, MSK, ChimeraTK Project <chimeratk-support@desy.de>
+// SPDX-License-Identifier: LGPL-3.0-or-later
+#pragma once
+
+#include <boost/enable_shared_from_this.hpp>
+#include <boost/shared_ptr.hpp>
+
+#include <mutex>
+
+namespace ChimeraTK {
+
+  class AsyncDomain : public boost::enable_shared_from_this<AsyncDomain> {
+   public:
+    virtual void sendException(const std::exception_ptr& e) noexcept = 0;
+    virtual ~AsyncDomain() = default;
+
+   protected:
+    // The mutex has to be recursive because an exception can occur within distribute, which is automatically triggering
+    // a sendException. This would deadlock with a regular mutex.
+    std::recursive_mutex _mutex;
+    // Everything in this class is protected by the mutex to the outside
+    // ATTENTION: The friends are not allowed to modify the _isActive flag, only to read it!!
+    bool _isActive{false};
```

Comment:
There is documentation missing. The trick is that the flag is protected by the mutex, and a public function must lock it. The friends must only be held by the AsyncDomain and their code is only called by the AsyncDomain while already holding the lock. You must not re-lock because it can lead to lock order inversion if another lock is involved somewhere in the backend.

Actually, this points to a bug: The mutex should not be recursive. Avoiding the lock order inversion is also the reason for the exception distribution thread in the AsyncDomainsContainer.

Reply:
I would suggest to call it `getIsActiveUnsafe()`, I think this is commonly used to mark functions that require special precautions when calling.

Reply:
I am not getting whether isActive is unsafe. How about ```unsafeGetIsActive()```? Or getIsActive_unsafe() (this bends the naming convention a little bit)

Reply:
Done. I used unsafeGetIsActive()